src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,888 @@
+                            }
+                            unfinished_splits.push(split);
+                        }
+                        BackfillState::SourceCachingUp(_) | BackfillState::Finished => {}
+                    }
+                    backfill_states.insert(split_id, backfill_state);
+                }
+            }
+        }
+        tracing::debug!(?backfill_states, "source backfill started");
+
+        // Return the ownership of `stream_source_core` to the source executor.
+        self.stream_source_core = core;
+
+        let source_chunk_reader = self
+            .build_stream_source_reader(&source_desc, unfinished_splits.clone())
+            .instrument_await("source_build_reader")
+            .await?;
+
+        fn select_strategy(_: &mut ()) -> PollNext {
+            futures::stream::PollNext::Left
+        }
+
+        // We choose "preferring upstream" strategy here, because:
+        // - When the upstream source's workload is high (i.e., Kafka has new incoming data), it just makes backfilling slower.
+        //   For chunks from upstream, they are simply dropped, so there's no much overhead.
+        //   So possibly this can also affect other running jobs less.
+        // - When the upstream Source's becomes less busy, SourceBackfill can begin to catch up.
+        let mut backfill_stream = select_with_strategy(
+            input.by_ref().map(Either::Left),
+            source_chunk_reader.map(Either::Right),
+            select_strategy,
+        );
+
+        type PausedReader = Option<impl Stream>;
+        let mut paused_reader: PausedReader = None;
+
+        macro_rules! pause_reader {
+            () => {
+                let (left, right) = backfill_stream.into_inner();
+                backfill_stream = select_with_strategy(
+                    left,
+                    futures::stream::pending().boxed().map(Either::Right),
+                    select_strategy,
+                );
+                // XXX: do we have to store the original reader? Can we simply rebuild the reader later?
+                paused_reader = Some(right);
+            };
+        }
+
+        // If the first barrier requires us to pause on startup, pause the stream.
+        if barrier.is_pause_on_startup() {
+            pause_reader!();
+        }
+
+        yield Message::Barrier(barrier);
+
+        let mut backfill_stage = BackfillStage {
+            states: backfill_states,
+            unfinished_splits,
+        };
+
+        if !self.backfill_finished(&backfill_stage.states).await? {
+            // We allow data to flow for `WAIT_BARRIER_MULTIPLE_TIMES` * `expected_barrier_latency_ms`
+            // milliseconds, considering some other latencies like network and cost in Meta.
+            let mut max_wait_barrier_time_ms = self.system_params.load().barrier_interval_ms()
+                as u128
+                * WAIT_BARRIER_MULTIPLE_TIMES;
```

Comment:
This was originally introduced as a backup for not making the epoch too large if there's something wrong with the meta service. But it appears to also make sense for backfilling. cc @yezizp2012 @chenzl25 what do you think?

Reply:
Agreed, I think we can keep it for backfilling.

Reply:
Agreed. BTW, to keep the same behavior as the previous one, maybe we should apply `max_wait_barrier_time_ms` to kafka backfill executor even after the backfill loop?

Reply:
after the backfill loop, Backfill only forwards msgs from Source (just like a Project), and Source contains this mechanism, so I think there's no need to add this in Backfill's forwarding stage.